### ISSUE TITLE
Regression: `docker compose config` does not output portable config anymore

### ISSUE THREAD

**Issue by jpic** [docker/compose#8914]
Previously, `docker-compose config` would output a portable (reusable) config, without names for volumes and networks that are global to docker.

Now, it enforces outputting local volume and network names at the top level networks and volumes hashes of the yaml output:

```
networks:
  default:
    name: ansible5rt8w0cmbuild_default
volumes:
  syncthing:
    name: ansible5rt8w0cmbuild_syncthing
```

Which breaks portability of the configuration generated from `docker-compose config` from a project to another anymore:

- default should not even be there
- volumes that are local to services should probably stay there at the service level

Is there any way to generate a portable config with the new docker compose?

Thank you

**Comment by jpic** [comment]
Worth mentioning another regression, getting `networks: { default: null }` in services as such:

```
services:
  some_service:
    networks:
      default: null
```

docker compose up won't complain about it, but the containers won't see each others.

**Comment by jpic** [comment]
Maybe related, one of the first things I fixed, was that instead of:

```
    volumes:
      - ./syncthing:/var/syncthing
      - ./data/media:/app/media
      - ./spooler:/spooler
      - ./log:/app/log
```

I'm now getting:

```
    - target: /app/log
      type: volume
      volume: {}
    - target: /app/media
      type: volume
      volume: {}
    - target: /tmp
      type: volume
      volume: {}
    - source: syncthing
      target: /var/syncthing
      type: volume
      volume: {}
```

Works but very verbose... I thought it was a feature but now I'm not sure ...

**Comment by ndeloof** [comment]
`docker-compose config` usage has never been clearly defined in compose V1.
For Compose V2, `config` command has been designed as a diagnostic solution to dump the actual, fully resolved model to be applied. This is why all "shot syntax" notation get extended, default values set, and implicit names resolved.

compose model portability is actually managed by the Compose Specification being adopted by various tools. What's your use case for this?

**Comment by jpic** [comment]
> compose model portability is actually managed by the Compose Specification being adopted by various tools. What's your use case for this?

My use case is building a single compose file from several files, inspect the target host and eventually hack it a bit and patch the file, and then deploy that. I can show you if you're curious (it's all open source and documented), but I don't want to distract this issue from what matters which I think is the following:

In compose v1, we could export a compose stack in a readable and portable format and deploy it say in another project or directory. It was never clearly defined, but it was working and used in production. Can you clarify if this feature is going to be dropped in v2? We can live without the "readable" part, since we generate our final compose files, but we rely on the portable part in many projects, from governmental to finance projects including NGOs (which I can also show).

Thank you very much

**Comment by ndeloof** [comment]
This "feature" has indeed been dropped. By the design of compose V2 being based on compose-go, we can't bump a compose.yaml model that would just render result from merging multiple yaml trees.

About the specific issue with `networks: default: null` I'm not sure what's wrong here, `null` being a valid yaml type

For your use case, what I don't understand is why you need this "pre-processing" of your compose files to "port" to many project. Why don't you just share the original yaml files and get projects to aggregate them on their own, + maybe an override file?

**Comment by jpic** [comment]
tl;dr I just need to know if you want me to keep reporting problems I run into with compose v2 config/convert or if the decision has been made to drop the feature in which case I can only recommend you document it so that people don't waste time trying to workaround all the problems. [Transitioning documentation in question](https://docs.docker.com/compose/cli-command/#transitioning-to-ga-for-compose-v2)

> About the specific issue with networks: default: null I'm not sure what's wrong here, null being a valid yaml type

It's perfectly valid yaml, but if you feed this yaml to `docker compose up` then your containers won't see each others. Sorry I haven't debugged why, I'm still picking up the pieces.

But anyway, why does it matter that it's perfectly valid yaml? After all, it's not re-usable anymore... Some would wonder why not just dump text or even csv...

Why does this command now provides information that we already have with `docker inspect` instead of generating portable configurations which I thought was the point of `convert` that `config` was an alias of? I think the use case was perfectly clear ...

> By the design of compose V2 being based on compose-go, we can't bump a compose.yaml model that would just render result from merging multiple yaml trees.

I was expecting that Go could do that, and that it would be trivial, if Go is not as powerful as Python it's fine and it's the reason I have not migrated to Go yet, we all know Python can do that because the Python docker-compose did that without problem. And it turns out my Ansible template filter is in Python so I might as well implement it and we'll be perfectly fine!

My question is all about the decision, yesterday I thought that I was helping to test the compose v2 transition that is advertised as "as smooth as possible" and I didn't know about the Go language limitations, so I've piled up some workaround code to keep using the config command... which does not seem not very useful anymore... What use case does this replacement feature that `docker inspect` doesn't already cover? That would be interesting to know! 

> Why don't you just share the original yaml files and get projects to aggregate them on their own, + maybe an override file?

The point of joining files is so that you can split your configuration in your repository, and then deploy a single file, and be able to call `docker compose` on the servers without having to be careful about passing proper `-f` flags!

Of course this is not useful if you're having one basic compose yml and then creating overrides manually on servers which is pretty cute too ;)

One docker-compose.yml + one override is not going to cut it for me, nor will it for many other users, probably the reason why the repeatable `-f` flag is there in `docker compose` in the first place!

> For your use case, what I don't understand is why you need this "pre-processing" of your compose files

I'm wary about going into details here, because it's not the point, I'm afraid to se my own requirements to be used as a justification to dropping a perfectly useful feature. But if you really want to know, it's all fully documented in the repository that contains the commits that this issue backlinks to.

>  to "port" to many project.

I'm not sharing compose files across different projects ... I'm not sure why I'd do that ... I'm pretty sure we're defining a different stack for each different project. I meant using the same workflow across different projects, the reason for that is because it's cost efficient. The point is to port a stack to different deployments, that's why you want to export a configuration. In my case [described here](https://yourlabs.org/posts/2020-02-08-bigsudo-extreme-devops-hacking-operations/#containers).

**Comment by mat007** [comment]
@jpic merging #8943 automatically closed this issue, but feel free to re-open it if it doesn’t fully address your need.
